### ISSUE TITLE
Weird dimension changing behavior

### ISSUE THREAD

**Issue by yoavartzi** [clab/dynet#239]
I noticed a weird behavior when adding two matrices. The dim() seems to be fine, as much as I understand how to read it, but the value() is a vector. The screenshot says it all. This is happening with the most recent version (just updated). 

![image](https://cloud.githubusercontent.com/assets/3289873/21941718/27d4a84c-d998-11e6-8f0d-0fbe57be503d.png)

**Comment by yoavg** [comment]
Confirmed. I will look into it in the following days. Thanks!

**Comment by yoavg** [comment]
This appears to be a C++ level thing. For some reason the ndims are set to 1 after the addition.

Code example:
```
  Expression x_ = parameter(cg,m.add_parameters({2,1}));
  cout << "x_.dim:" << x_.dim() << endl;
  cout << "(x_ + x_).dim:" << (x_ + x_).dim() << endl;
  Dim d = (x_ + x_).dim();
  cout << "(x_ + x_).nrows,ncols,ndims:" << d.rows() << " " << d.cols() << " " << d.ndims() << endl;
```

**Comment by neubig** [comment]
The C++ version of the bug should be fixed by https://github.com/clab/dynet/commit/7f45b7e6b1dd85f510f7e6ce362dfd1e8942b52a

**Comment by neubig** [comment]
Looks like the Python is still broken though. Any ideas @yoavg ?

**Comment by pmichel31415** [comment]
I think I corrected this issue in #340 : 

Basically what `expr.dim()` did was returning (nb of elements,nb of rows, nb of cols, nb of batches) so here (2,2,1,1)

The new behavior in my pull request is returning (dim_1,dim_2,...,dim_n,batch_size) so this would correspond to (2,1,1) here.

Maybe it would be better not to include the batch size if it's equal to 1, what are your thoughts?

**Comment by neubig** [comment]
I think it should probably return a tuple `((2,1), 1)`, which is probably what you did?

**Comment by pmichel31415** [comment]
Aah that makes more sense, I'll do that.

**Comment by neubig** [comment]
Fixed by #340?
